src/OpenTelemetry/Trace/Export/BatchingActivityProcessor.cs

```diff
@@ -0,0 +1,244 @@
+        private readonly ConcurrentQueue<Activity> exportQueue;
+        private readonly int maxQueueSize;
+        private readonly int maxExportBatchSize;
+        private readonly TimeSpan scheduleDelay;
+        private readonly ActivityExporter exporter;
+        private CancellationTokenSource cts;
+        private volatile int currentQueueSize;
+        private bool stopping = false;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchingActivityProcessor"/> class with default parameters:
+        /// <list type="bullet">
+        /// <item>
+        /// <description>maxQueueSize = 2048,</description>
+        /// </item>
+        /// <item>
+        /// <description>scheduleDelay = 5 sec,</description>
+        /// </item>
+        /// <item>
+        /// <description>maxExportBatchSize = 512</description>
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="exporter">Exporter instance.</param>
+        public BatchingActivityProcessor(ActivityExporter exporter)
+            : this(exporter, DefaultMaxQueueSize, DefaultScheduleDelay, DefaultMaxExportBatchSize)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchingActivityProcessor"/> class with custom settings.
+        /// </summary>
+        /// <param name="exporter">Exporter instance.</param>
+        /// <param name="maxQueueSize">Maximum queue size. After the size is reached activities are dropped by processor.</param>
+        /// <param name="scheduleDelay">The delay between two consecutive exports.</param>
+        /// <param name="maxExportBatchSize">The maximum batch size of every export. It must be smaller or equal to maxQueueSize.</param>
+        public BatchingActivityProcessor(ActivityExporter exporter, int maxQueueSize, TimeSpan scheduleDelay, int maxExportBatchSize)
+        {
+            if (maxQueueSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxQueueSize));
+            }
+
+            if (maxExportBatchSize <= 0 || maxExportBatchSize > maxQueueSize)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxExportBatchSize));
+            }
+
+            this.exporter = exporter ?? throw new ArgumentNullException(nameof(exporter));
+            this.maxQueueSize = maxQueueSize;
+            this.scheduleDelay = scheduleDelay;
+            this.maxExportBatchSize = maxExportBatchSize;
+
+            this.cts = new CancellationTokenSource();
+            this.exportQueue = new ConcurrentQueue<Activity>();
+
+            // worker task that will last for lifetime of processor.
+            // No need to specify long running - it is useless if any async calls are made internally.
+            // Threads are also useless as exporter tasks run in thread pool threads.
+            Task.Factory.StartNew(s => this.Worker((CancellationToken)s), this.cts.Token);
```

Comment:
No reason to hold up this PR, though I am curious about this usage of `Task.Factory.StartNew`. This is usually used for more complex scenarios where it's necessary to configure how the task is scheduled/executed. The usage may be safe here though it can be a source of [problems](https://blog.stephencleary.com/2013/08/startnew-is-dangerous.html).

Looks like this got introduced in #286, though it doesn't seem clear to me why `Task.Run` was changed to `Task.Factory.StartNew`.

Reply:
Thanks for the link @alanwest, I have read both Stephen Clearly and Stephen Toub's posts on `Task.Run` vs `Task.Factory.Start` on many occasions but still see it come up often. A Roslyn analyzer might be useful to check you actually _need_ it :smile:.

I agree `Task.Run` is likely the better option here as we do no need the extra options `Task.Factory.Start` offers. I didn't think about it when I ported the this from the span version.